glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,227 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+//! Module for registering shared types for Rust types.
+
+use crate::translate::*;
+
+pub trait RefCounted: Clone + Sized + 'static {
+    /// The inner type
+    type InnerType;
```

Comment:
Why as a associated type and not as a generic parameter (i.e. `pub trait RefCounted<T: SharedType>: Clone + Sized + 'static`)?

Reply:
Just because this way it turned out to be easier to implement the `SharedType` trait inside the macro. I'm OK to change it of course if we have a strong preference for the generic parameter :)

Reply:
How is it more difficult with a generic parameter? I think I would prefer that because it maps directly to how `Arc` and `Rc` are defined, unless it makes things much more difficult.

But in either case, you can make `type SharedType: RefCounted<InnerType=Self>` or `type SharedType: RefCounted<Self>`.

Reply:
I'm sorry but I'm getting a bit confused by:
> But in either case, you can make type SharedType: RefCounted<InnerType=Self> or type SharedType: RefCounted<Self>.

The `RefCounted` trait is the one that is implemented by `Arc<T>/Rc<T>`, and `InnerType`, whether it is a generic parameter or an associated type of `RefCounted`, should not implement any particular trait, just be the same of `T`. What am I missing?

Actually, now that I think more about that, I believe that the "associated type" version makes probably more sense.
If we implement `RefCounted` as a generic parameter:
```
pub trait RefCounted<T>: Clone + Sized + 'static {
    unsafe fn ref_(this: *const T) -> *const T;
    ...
}
```
we are potentially allowing `Arc/Rc<V>` to implement it in different ways (meaning the parameter `V` not necessarily the same as `T`), while I think it makes sense to enforce those to be the same.
What do you think?

Reply:
See my comment here at the code in question https://github.com/gtk-rs/gtk-rs/pull/452/files#r613899233

I don't think either of the implementations makes a practical difference if that is done at that place, so keeping it as associated type should also be fine.

Reply:
Oh, now I see my confusion. `type RefCountedType` is going to be something like `Arc<FooInner>`, so the things are actually not matching up in that way and that can't work. The outer and inner types are different.

So yes, this is actually all fine. Sorry!